table/block_based_table_reader.cc

```diff
@@ -126,22 +126,44 @@ Slice GetCacheKeyFromOffset(const char* cache_key_prefix,
 Cache::Handle* GetEntryFromCache(Cache* block_cache, const Slice& key,
                                  Tickers block_cache_miss_ticker,
                                  Tickers block_cache_hit_ticker,
-                                 Statistics* statistics) {
+                                 Statistics* statistics,
+                                 GetContext* get_context) {
   auto cache_handle = block_cache->Lookup(key, statistics);
   if (cache_handle != nullptr) {
     PERF_COUNTER_ADD(block_cache_hit_count, 1);
-    // overall cache hit
-    RecordTick(statistics, BLOCK_CACHE_HIT);
-    // total bytes read from cache
-    RecordTick(statistics, BLOCK_CACHE_BYTES_READ,
-               block_cache->GetUsage(cache_handle));
-    // block-type specific cache hit
-    RecordTick(statistics, block_cache_hit_ticker);
+    if (get_context != nullptr) {
+      // overall cache hit
+      get_context->RecordCounters(BLOCK_CACHE_HIT, 1);
+      // total bytes read from cache
+      get_context->RecordCounters(BLOCK_CACHE_BYTES_READ,
+                                   block_cache->GetUsage(cache_handle));
+      // block-type specific cache hit
+      if (block_cache_hit_ticker == BLOCK_CACHE_DATA_HIT) {
+        get_context->RecordCounters(BLOCK_CACHE_DATA_HIT, 1);
+      } else {
+        RecordTick(statistics, block_cache_hit_ticker);
+      }
+    } else {
+      // overall cache hit
+      RecordTick(statistics, BLOCK_CACHE_HIT);
+      // total bytes read from cache
+      RecordTick(statistics, BLOCK_CACHE_BYTES_READ,
+                 block_cache->GetUsage(cache_handle));
+      RecordTick(statistics, block_cache_hit_ticker);
+    }
   } else {
     // overall cache miss
-    RecordTick(statistics, BLOCK_CACHE_MISS);
+    if (get_context != nullptr) {
+      get_context->RecordCounters(BLOCK_CACHE_MISS, 1);
+    } else {
+      RecordTick(statistics, BLOCK_CACHE_MISS);
+    }
     // block-type specific cache miss
-    RecordTick(statistics, block_cache_miss_ticker);
+    if (block_cache_miss_ticker == BLOCK_CACHE_DATA_MISS &&
```

Comment:
also here you may be able to call RecordCounters regardless of whether it's BLOCK_CACHE_DATA_MISS